README.md
# RTSPBrute

[![pipeline status](https://gitlab.com/woolf/RTSPbrute/badges/master/pipeline.svg)](https://gitlab.com/woolf/RTSPbrute/-/commits/master)
[![coverage report](https://gitlab.com/woolf/RTSPbrute/badges/master/coverage.svg)](https://gitlab.com/woolf/RTSPbrute/-/commits/master)

<p align="center">
   <a href="https://asciinema.org/a/353291" target="_blank"><img src="https://asciinema.org/a/353291.svg" /></a>
</p>

> Inspired by [Cameradar](https://github.com/Ullaakut/cameradar)

> Forked from https://gitlab.com/woolf/RTSPbrute

## DISCLAIMER
This software is provided for research and ethical uses only!

## Features

- **Find accessible RTSP streams** on any target
- Brute-force **stream routes**
- Brute-force **credentials**
- **Make screenshots** on accessible streams
- Generate **user-friendly report** of the results:
  - `.txt` file with each found stream on new line
  - `.html` file with screenshot of each found stream

### Report files

- `result.txt`: Each target is on a new line. Import to VLC: change extension to `.m3u` and open in VLC
- `index.html`: Click on the screenshot to copy its link

## Installation

### Requirements

- `python` (> `3.6`)
- `av`
- `Pillow`
- `rich`

Install with `pip` or your favorite PyPi package manager.

```
pip install rtspbrute
```

## CLI

```
USAGE
    $ rtspbrute -t TARGETS [-p PORTS [PORTS ...]] [-r ROUTES] [-c CREDENTIALS]
                [-ct N] [-bt N] [-st N] [-T TIMEOUT] [-d] [-h]

ARGUMENTS
    -h, --help                     show this help message and exit
    -t, --targets TARGETS          the targets on which to scan for open RTSP streams
    -p, --ports PORTS [PORTS ...]  the ports on which to search for RTSP streams
    -r, --routes ROUTES            the path on which to load a custom routes
    -c, --credentials CREDENTIALS  the path on which to load a custom credentials
    -ct, --check-threads N         the number of threads to brute-force the routes
    -bt, --brute-threads N         the number of threads to brute-force the credentials
    -st, --screenshot-threads N    the number of threads to screenshot the streams
    -T, --timeout TIMEOUT          the timeout to use for sockets
    -d, --debug                    enable the debug logs

EXAMPLES
    $ rtspbrute -h
    $ rtspbrute -t hosts.txt -p 554 5554 8554 -d
    $ rtspbrute -t ips.txt -r routes.txt -c combinations.txt
    $ rtspbrute -t targets.txt -st 10 -T 10
```

### **"argument"** (`default_value`):

- **"-t, --targets"** (_No default value_): Set the path to the input file. The file can contain IPs, IP ranges and CIDRs. Each one of them should be on a separate line, e.g.:

```
0.0.0.0
192.168.100.1-192.168.254.1
192.17.0.0/16
```

- **"-p, --ports"** (`554`): Set custom ports, e.g.: `-p 554 5554 8554`
- **"-r, --routes"** (`routes.txt`): Set custom path to the file with routes. Each route should start with `/` and be on a separate line, e.g.:

```
/1
/11
/h264
```

- **"-c, --credentials"** (`credentials.txt`): Set custom path to the file with credentials. Each combination should contain `:` and be on a separate line, e.g.:

```
admin:admin
user:user
```

- **"-ct, --check-threads"** (`500`): Set custom number of threads to brute-force the routes
- **"-bt, --brute-threads"** (`200`): Set custom number of threads to brute-force the credentials
- **"-st, --screenshot-threads"** (`20`): Set custom number of threads to screenshot the streams. Smaller number leads to more successful screenshots: when there's too much threads PyAV will throw errors and wouldn't connect to target.
- **"-T, --timeout"** (`2`): Set custom timeout value for socket connections
- **"-d, --debug"** (`False`): Enable debug logging to `debug.log` file


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/),
and this project adheres to [Semantic Versioning](https://semver.org/).

## [Unreleased]

## [1.0.4] - 2021-05-17
### Fixed
- Issue #2.

## [1.0.3] - 2020-08-20
### Added
- Setting the soft limit of NOFILE to maximum on Linux.
- The total number of images, and a tip to top of the HTML file.

### Changed
- Debug logging to ignore timeout errors.

## [1.0.2] - 2020-08-14
### Fixed
- Creation of a file on the start.

## [1.0.1] - 2020-08-14
### Added
- Routes from cameradar.

### Changed
- Minimal Python version to 3.6.

## [1.0.0] - 2020-08-14
### Added
- First official release.

[Unreleased]: https://gitlab.com/woolf/RTSPbrute/-/compare/1.0.4...master
[1.0.4]: https://gitlab.com/woolf/RTSPbrute/-/compare/1.0.3...1.0.4
[1.0.3]: https://gitlab.com/woolf/RTSPbrute/-/compare/1.0.2...1.0.3
[1.0.2]: https://gitlab.com/woolf/RTSPbrute/-/compare/1.0.1...1.0.2
[1.0.1]: https://gitlab.com/woolf/RTSPbrute/-/compare/1.0.0...1.0.1
[1.0.0]: https://gitlab.com/woolf/RTSPbrute/-/commits/1.0.0


rtspbrute/modules/cli/input.py
import argparse
from pathlib import Path
from typing import Any

from rtspbrute import DEFAULT_CREDENTIALS, DEFAULT_ROUTES, __version__


class CustomHelpFormatter(argparse.HelpFormatter):
    def __init__(self, prog):
        super().__init__(prog, max_help_position=40, width=99)

    def _format_action_invocation(self, action):
        if not action.option_strings or action.nargs == 0:
            return super()._format_action_invocation(action)
        default = self._get_default_metavar_for_optional(action)
        args_string = self._format_args(action, default)
        return ", ".join(action.option_strings) + " " + args_string


def file_path(value: Any):
    if Path(value).exists():
        return Path(value)
    else:
        raise argparse.ArgumentTypeError(f"{value} is not a valid path")


def port(value: Any):
    if int(value) in range(65536):
        return int(value)
    else:
        raise argparse.ArgumentTypeError(f"{value} is not a valid port")


fmt = lambda prog: CustomHelpFormatter(prog)
parser = argparse.ArgumentParser(
    prog="rtspbrute",
    description="Tool for RTSP that brute-forces routes and credentials, makes screenshots!",
    formatter_class=fmt,
)
parser.add_argument(
    "-t",
    "--targets",
    type=file_path,
    required=True,
    help="the targets on which to scan for open RTSP streams",
)
parser.add_argument(
    "-p",
    "--ports",
    nargs="+",
    default=[554],
    type=port,
    help="the ports on which to search for RTSP streams",
)
parser.add_argument(
    "-r",
    "--routes",
    type=file_path,
    default=DEFAULT_ROUTES,
    help="the path on which to load a custom routes",
)
parser.add_argument(
    "-c",
    "--credentials",
    type=file_path,
    default=DEFAULT_CREDENTIALS,
    help="the path on which to load a custom credentials",
)
parser.add_argument(
    "-ct",
    "--check-threads",
    default=1000,
    type=int,
    help="the number of threads to brute-force the routes",
    metavar="N",
)
parser.add_argument(
    "-bt",
    "--brute-threads",
    default=200,
    type=int,
    help="the number of threads to brute-force the credentials",
    metavar="N",
)
parser.add_argument(
    "-st",
    "--screenshot-threads",
    default=20,
    type=int,
    help="the number of threads to screenshot the streams",
    metavar="N",
)
parser.add_argument(
    "-T", "--timeout", default=2, type=int, help="the timeout to use for sockets"
)
parser.add_argument("-d", "--debug", action="store_true", help="enable the debug logs")
parser.add_argument("--version", action="version", version=f"%(prog)s {__version__}")
